### PR TITLE
Cluster scoped DataVolume/PVC namespace transfer API

### PR DESCRIPTION
**Pull request by mhenriks**
**What this PR does / why we need it**:

Move DataVolumes/PVCs around like this:

```yaml
apiVersion: cdi.kubevirt.io/v1beta1
kind: ObjectTransfer
metadata:
  name: t1
spec:
  source:
    kind: DataVolume # can also be PersistentVolumeClaim
    namespace: default
    name: source
  target:
    namespace: test
```

The transfer controller will wait until the DV/PVC is populated and not in use before initiating (no copy) transfer.

To wait for the transfer to complete, run something like this:

```bash
kubectl wait objecttransfer t1 --for condition=Complete
```

ObjectTransfer specs are immutable.  Webhook will validate that target does not exist.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Cluster scoped DataVolume/PVC namespace transfer API
```

### CODE REVIEW

**Comment by awels** [comment]
Looks like some gofmt issues in travis

File: cluster-sync/clean.sh [hunk 1]
```diff
@@ -20,6 +20,13 @@ set +e
 _kubectl patch cdi cdi --type=json -p '[{ "op": "remove", "path": "/metadata/finalizers" }]'
 set -e
 
+_kubectl get ot --all-namespaces -o=custom-columns=NAME:.metadata.name,NAMESPACE:.metadata.namespace,FINALIZERS:.metadata.finalizers --no-headers | grep objectTransfer | while read p; do
+    arr=($p)
+    name="${arr[0]}"
+    namespace="${arr[1]}"
+    _kubectl patch ot $name -n $namespace --type=json -p '[{ "op": "remove", "path": "/metadata/finalizers" }]'
+done
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply cluster-sync/clean.sh:28]: Yeah, the ObjectTransfer has a finalizer when running.  So if a test fails in that time, have to clean up

File: pkg/apis/core/v1beta1/types_transfer.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	Name *string `json:"name,omitempty"`
+}
+
+// ObjectTransferSpec specifies the source/target of the transfer
+type ObjectTransferSpec struct {
+	Source TransferSource `json:"source"`
+
+	Target TransferTarget `json:"target"`
+
+	ParentName *string `json:"parentName,omitempty"`
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/apis/core/v1beta1/types_transfer.go:71]: Transferring a DataVolume creates an ObjectTransfer internally for the PVC.  But I want to reconcile the DataVolumeTransfer whenever the PVC transfer is updated.  So added this field and it is checked in the watch code.

Check this out:  https://github.com/mhenriks/containerized-data-importer/blob/transfer-api/pkg/controller/transfer/watch.go#L77-L100

File: pkg/controller/transfer/controller.go [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+	Scheme   *runtime.Scheme
+	Log      logr.Logger
+}
+
+func getTransferTargetName(ot *cdiv1.ObjectTransfer) string {
+	if ot.Spec.Target.Name != nil {
+		return *ot.Spec.Target.Name
+	}
+
+	return ot.Spec.Source.Name
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/controller/transfer/controller.go:75]: Pretty sure it's used for more than the logger.  And the goal is to not have to deal with nil or "".  The target must have a name.  It is either the same as the source or explicitly specified.

File: cluster-sync/clean.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (awels) [reply cluster-sync/clean.sh:28]: If I read the code correctly, we should be able to simply delete the transfer objects, the controller takes care of removing the finalizers if the deletionTimestamp != nil. Of course if the controller is not running, that would be a problem.

File: pkg/apis/core/v1beta1/types_transfer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (awels) [reply pkg/apis/core/v1beta1/types_transfer.go:71]: Gotcha

File: cluster-sync/clean.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (mhenriks) [reply cluster-sync/clean.sh:28]: Only if the transfer is pending/complete

File: pkg/controller/transfer/pvc.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+		return 0, nil
+	}
+
+	pods, err := cdicontroller.GetPodsUsingPVCs(h.reconciler.Client, pvc.Namespace, sets.NewString(pvc.Name), false)
+	if err != nil {
+		return 0, h.reconciler.setCompleteConditionError(ot, err)
+	}
+
+	if len(pods) > 0 {
```
Comment (aglitke) [inline pkg/controller/transfer/pvc.go:49]: Just wondering if there is a race condition here between transferring and consuming a PVC?  What happens if we rebind the PVC while it's being used by a Pod in the source ns?

File: pkg/controller/transfer/pvc.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+			if err := h.reconciler.updateResource(ot, pv); err != nil {
+				return 0, h.reconciler.setCompleteConditionError(ot, err)
+			}
+
+			return 0, h.reconciler.setCompleteConditionRunning(ot)
+		}
+
+		if source.DeletionTimestamp == nil {
+			if err := h.reconciler.Client.Delete(context.TODO(), source); err != nil {
+				return 0, h.reconciler.setCompleteConditionError(ot, err)
```
Comment (aglitke) [inline pkg/controller/transfer/pvc.go:115]: Do you need to rollback the PVC reclaim policy here?

Comment (aglitke) [reply pkg/controller/transfer/pvc.go:115]: Nevermind, Maybe I'm just confused about eventual consistency.  This loop will continue trying each time until it gets all the way through.

**Review by aglitke — COMMENTED** [review]
Do we need to document this API yet or is this really just for the clone controller's internal use?

File: pkg/controller/transfer/pvc.go (hunk 4 above)
Comment (mhenriks) [reply pkg/controller/transfer/pvc.go:49]: That race cannot be eliminated as far as I know without changing kubelet.  As to what happens either the pod will be pending or the PVC will no get deleted.  The transfer controller will keep trying.

**Comment by mhenriks** [comment]
This is meant to be an internal API.  Namespace scoped will be documented.

**Pull request merged**